### PR TITLE
refactor: remove traceWriter singleton from span-data.ts

### PR DESCRIPTION
**Pull request by kjin**
In preparation for a fix for #1025, this PR removes the references to the `traceWriter` singleton used in `SpanData` classes. It was used for two things:
* Publishing spans
* Getting span label limits

Now:
* Spans can be _enabled for publish_ by having a `TraceWriter` instance passed to them. This is so that in the future, we can have some spans that are "sampled" but not published, in order to address the issue mentioned above.
* Span label limits have been moved to `Tracer` config options instead, and passed in through the corresponding `SpanData` constructor.

### CODE REVIEW

File: src/trace-api.ts [hunk 1]
```diff
@@ -241,7 +242,7 @@ export class StackdriverTracer implements Tracer {
       rootContext = UNTRACED_ROOT_SPAN;
     } else {
       // Create a new root span, and invoke fn with it.
-      rootContext = new RootSpanData(
+      const rootSpan = new RootSpanData(
```
Comment (ofrobots) [inline src/trace-api.ts]: Why did you change this and add a new variable? Assigning to `rootContext` here should still work right?

File: src/span-data.ts [hunk 2]
```diff
@@ -159,14 +179,16 @@ export class RootSpanData extends BaseSpanData implements RootSpan {
...
     super.endSpan(timestamp);
-    traceWriter.get().writeTrace(this.trace);
-    this.children.forEach(child => {
-      if (!child.span.endTime) {
-        // Child hasn't ended yet.
-        // Inform the child that it needs to self-publish.
-        child.shouldSelfPublish = true;
-      }
-    });
+    if (this.traceWriter) {
```
Comment (ofrobots) [inline src/span-data.ts]: Consider the scenario where this code is reached, and `this.traceWriter` is not defined. Is that an error, or is that something very expected – and hence, silently ignorable, as being done here?

Comment (kjin) [reply src/span-data.ts]: Right now there's no code path that causes `this.traceWriter` not to be defined, but this code path will exist in a forthcoming PR, where we would opt to not publish a trace, but still propagate context.

File: src/trace-api.ts (hunk 1 above)
Comment (kjin) [reply src/trace-api.ts]: No, since `rootContext` is not a `RootSpanData`, and won't have a known `assignTraceWriter` field.

File: src/span-data.ts (hunk 2 above)
Comment (ofrobots) [reply src/span-data.ts]: IMO it would be better for the code to be self-documenting. `if (this.shouldPublish) { /* publish code */ }` would be more obviously readable. Using the validness of `this.traceWriter` to accomplish that, but it doesn't clarify whether it is possible that some bug may cause failure to call `assignTraceWriter`, causing traces to be silently dropped.

IOW, my question is: did we really need to get rid of the singleton? Wouldn't it make sense to keep it a singleton, but add boolean state to each span on whether it should be published?

Comment (kjin) [reply src/span-data.ts]: > did we really need to get rid of the singleton?

I've decided that we don't need to get rid of it in order to address the issue at hand; I'll open a different PR with an alternative approach. That being said, I still think it is better not to depend on the singleton, and have the `TraceWriter` passed in -- it makes the code more testable and cleaner IMO. It should be considered a testable behavior of `Tracer` that it always calls `assignTraceWriter` so I don't know that a bug could crop up as long as we keep our tests intact. (I'll admit though that `assignTraceWriter` might not be the best name.)

Given that the change is ultimately inconsequential to users, though, I'd be OK with closing this PR.

**Comment by kjin** [comment]
Superceded by work in #1033.
